include/zephyr/bin_descriptors.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2023 Yonatan Schachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZEPHYR_BIN_DESCRIPTORS_H_
+#define ZEPHYR_INCLUDE_ZEPHYR_BIN_DESCRIPTORS_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define BIN_DESCRIPTOR_MAGIC 0x2c31b9b9
+
+#ifndef _LINKER
+/**
+ * @brief A union for holding the value of the binary descriptor, which
+ * can be either an integer or a string.
+ */
+union bin_descriptor_value {
+	/** Value as an integer */
+	uint32_t integer;
+	/** Value as a pointer to a char */
+	const char *str;
+};
```

Comment:
I'd also suggest reworking this to `type` / `len` / `data`... this way knowing the start address only will permit you "walk" through all the items.

If non-string / BLOB types (e.g: `uint32_t`, pointers, etc...) are really expected, then either special handling or alignment becomes important.

```suggestion
struct bin_descriptor {
	uint32_t type; /* e.g: BIN_DESC_TYPE_APP_VERSION_STRING */
	uint32_t len; /* length of data, below */
	uint8_t data[];
};
```